api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto

```diff
@@ -81,7 +92,14 @@ message RedisProxy {
 
   message RouteConfig {
     repeated Route routes = 1 [(gogoproto.nullable) = false];
+
+    // Catch-all route for commands that doesn't match any of the routes.
+    RouteAction catch_all = 2;
```

Comment:
Instead of having an explicit catch all, can we just require that the routes list above has at least 1 element in it?

Reply:
I did it as an explicit catch all mostly because the `RouteMatch` doesn't really make sense for a catch all route. Unless we want to use a sentinel value such as `*` as a prefix? Or the route could have many match variants.

```proto
message Route {
  oneof match {
    Prefix prefix = 1 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
    CatchAll catch_all = 2;
  }

  // Route request to some upstream cluster.
  RouteAction route = 2 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
}
```

WDYT?

Reply:
Personally, I think I would still require a `RouteMatch`, but inside I would make the match type a oneof, which currently has 2 values, prefix and "any_match", which is a bool that can only be set to true (search for this in the API, we do this elsewhere). This would allow us flexibility in the future to add regex, etc. Then for users that just want to do what they do today, they would do a single route match, with an any_match. WDYT?

Reply:
Yes definitely make sense to have a oneof type to support regex, or any other type of routes. However, since there can only be a 0..1 catch all route, I thought it would make sense to have it as an optional route at the RouteConfiguration level. Otherwise, I'm not sure how to handle if there's multiple `any_match` routes, do we forward only to the first one?

Reply:
Yes, I would just do FIFO matching like we do for HTTP. I think what I have proposed is most consistent with what we do already for HTTP FWIW.

Reply:
TIL. I assume the data structure for http routes are a little different (Trie?), here I was naively storing the prefix routes in an ordered set, then dispatching to the catch-all if it's defined. 

I'm all in for making it consistent, I'll give it a shot 👍I initially tried to avoid any app level check and thought [mcrouter](https://github.com/facebook/mcrouter/wiki/Prefix-routing-setup)'s implementation made sense to do it as such.

Reply:
Well, if you were planning on doing a hash lookup, I think that's reasonable, but then we can't support regex anyway. If you think we will only ever support a prefix lookup, I think that's probably fine. What would be most useful here?

Reply:
I think prefix and regex are the 2 candidates I can think of. Although I don't personally have a use case for regexes, but I can see why it could be useful.

Reply:
I think I would decide on whether you think you would favor performance (hash lookup, less matching options) or flexibility (FIFO). For Redis, favoring performance seems reasonable here, so I'm fine either way. I would just document how matches work if we are going to go for a hashing system.

Reply:
I've been thinking about this overnight, and I currently prefer this implementation for a few reasons, let me know if you agree with my reasoning 😄 

1. The semantic for any matches is not used for the path or prefix matching, but for the [tap service](https://github.com/maximebedard/envoy/blame/master/api/envoy/service/tap/v2alpha/common.proto#L51-L52) match predicates. I believe it provides a much more complex boolean logic API than what we really need here.

2. As for HTTP, a rule with `path: /` is effectively a catch all. Here, we could do the same with using an any_match parameter, or using a sentinel such as `prefix: *`, but given that we can only have 0..1 value, I thought it was sane to embed this behavior directly in the protocol and not do any app-level check.

If I were to use an `any_match` parameter, I think I would change the protocol as such:

```proto
message PrefixMatch {
  string prefix = 1 [(validate.rules).string.min_bytes = 1];
  bool case_insensitive = 2;
}

// separate message for catch-all since case_insensitive does not really make sense for this variant.
message CatchAllMatch {
  bool any_match;
}

// TODO: regex match

message Route {
  oneof match {
    PrefixMatch prefix = 1 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
    CatchAllMatch catch_all = 2;
    RegexMatch regex = 3;
  }

  RouteAction route = 4 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
}
```

Given that the catch-all scenario is sort-of a lack of anything that matches, I think it makes sense to avoid defining a "match" expression for it.

Let me know what you think, I'll try to implement it as such today too to see what the implementation details would look like!

Reply:
Yeah I think given the fact that we want to support hash matching and not FIFO, this makes sense. Can you update the docs to make it clear that hashing is being used? This will also make it clear that we cannot support regex in the future (although even in the current implementation I'm not exactly sure how you would handle overlapping prefixes without more work?)

Reply:
In the current implementation, I'm using an ordered set to stored the prefix routes. Overlapping prefixes will be ignored, and it will look for the longest match first.

Reply:
SGTM, I would just fully document the expected behavior and I will review the code after!